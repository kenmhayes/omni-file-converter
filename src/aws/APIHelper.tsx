import { API } from 'aws-amplify';

const SESSION_API_NAME = 'SessionGatewayAPI';

export interface CreateSessionRequest {
  conversionRequests: FileConversionInput[];
}

export interface FileConversionInput {
  fileName: string;
  originalFileFormat: string;
  convertedFileFormat: string;
}

interface CreateSessionResponse {
  sessionId: string;
  fileNameToId: { [key: string]: string };
}

export interface GetSessionResponse {
  expirationTime: number;
  conversions: ConversionOutput[];
}

export interface ConversionOutput {
  id: string;
  fileName: string;
  conversionStatus: string;
  originalFileFormat: string;
  convertedFileFormat: string;
}

// eslint-disable-next-line max-len
export const createSession = async (request: CreateSessionRequest): Promise<CreateSessionResponse> => API.post(
  SESSION_API_NAME,
  '/createsession',
  {
    body: request,
  },
);

export const getSession = async (sessionId: string): Promise<GetSessionResponse> => API.get(
  SESSION_API_NAME,
  '/getsession',
  {
    queryStringParameters: {
      sessionid: sessionId,
    },
  },
);
