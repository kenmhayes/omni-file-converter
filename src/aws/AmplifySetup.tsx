import { Amplify, Auth, Storage } from 'aws-amplify';

function AmplifySetup() {
  /**
 * Initialize Cognito authentication
 */
  Amplify.configure({
    Auth: {
      identityPoolId: process.env.REACT_APP_IDENTITY_POOL_ID,
      region: process.env.REACT_APP_AWS_REGION,
    },
    Storage: {
      AWSS3: {
        bucket: process.env.REACT_APP_S3_BUCKET,
        region: process.env.REACT_APP_AWS_REGION,
      },
    },
  });

  Auth.configure();

  const getCredentials = async () => {
    await Auth.currentCredentials();
  };

  getCredentials();

  // Configure S3 storage
  Storage.configure({
    customPrefix: {
      public: '',
    },
  });
}

export default AmplifySetup;
