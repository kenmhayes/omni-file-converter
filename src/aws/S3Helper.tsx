import { Storage } from 'aws-amplify';

export const getS3ObjectUrl = async (objectKey: string): Promise<string> => Storage.get(
  `${process.env.REACT_APP_S3_CONVERTED_PREFIX}${objectKey}`,
);

export const putS3Object = async (
  objectKey: string,
  convertType: string,
  file: File,
): Promise<void> => {
  await Storage.put(`${process.env.REACT_APP_S3_ORIGINAL_PREFIX}${convertType}/${objectKey}`, file);
};
