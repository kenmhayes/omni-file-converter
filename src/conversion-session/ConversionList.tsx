import React, { useState } from 'react';
import {
  Anchor, Button, Card, Table,
} from 'react-bootstrap';
import { ConversionOutput } from '../aws/APIHelper';
import { getS3ObjectUrl } from '../aws/S3Helper';
import READABLE_CONVERSION_STATUSES from '../constants/ConversionStatuses';

export interface ConversionListProps {
  conversions: ConversionOutput[];
}

interface ConversionListItemProps {
  conversion: ConversionOutput;
}

function ConversionListItem(props: ConversionListItemProps) {
  const { conversion } = props;
  const [downloadUrl, setDownloadUrl] = useState<string>('');

  const onButtonClick = async () => {
    const objectKey = `${conversion.id}.${conversion.convertedFileFormat}`;

    const objectDownloadUrl = await getS3ObjectUrl(objectKey);
    setDownloadUrl(objectDownloadUrl);
  };

  return (
    <tr>
      <td>{conversion.fileName}</td>
      <td>{conversion.convertedFileFormat}</td>
      <td>{READABLE_CONVERSION_STATUSES[conversion.conversionStatus]}</td>
      <td>
        {
            downloadUrl !== ''
              ? <Anchor href={downloadUrl}>Download</Anchor>
              : <Button onClick={onButtonClick}>Get Download Link</Button>
        }
      </td>
    </tr>
  );
}

function ConversionList(props: ConversionListProps) {
  const { conversions } = props;

  return (
    <div className="container">
      <Card>
        <Card.Header>
          <h4>Your converted files:</h4>
        </Card.Header>
        <Card.Body>
          <Table>
            <thead>
              <tr>
                <th>File Name</th>
                <th>Converted File Type</th>
                <th>Status</th>
                <th>Download Link</th>
              </tr>
            </thead>
            <tbody>
              {conversions.map((conversion: ConversionOutput) => (
                <ConversionListItem conversion={conversion} key={conversion.id} />
              ))}
            </tbody>
          </Table>
        </Card.Body>
      </Card>
    </div>
  );
}

export default ConversionList;
