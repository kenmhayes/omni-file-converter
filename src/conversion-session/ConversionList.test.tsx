import React from 'react';
import { act, create } from 'react-test-renderer';
import { fireEvent, render } from '@testing-library/react';
import ConversionList from './ConversionList';
import { ConversionOutput } from '../aws/APIHelper';
import * as S3Helper from '../aws/S3Helper';

const s3Spy = jest.spyOn(S3Helper, 'getS3ObjectUrl').mockResolvedValue("test");

describe('ConversionList', () => {
    const conversions: ConversionOutput[] = [
        {
            fileName: 'test1.jpg',
            id: '1234',
            conversionStatus: 'COMPLETED',
            originalFileFormat: 'jpg',
            convertedFileFormat: 'png'
        }
    ];


    it('renders a conversion list', () => {
        const component = create(
          <ConversionList conversions={conversions}/>
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('gets a download url when the button is clicked', async () => {    
        const component = <ConversionList conversions={conversions}/>;
        const { queryByRole, getByRole, findByText } = render(component);
    
        const downloadButtonRole = "button";
        expect(queryByRole(downloadButtonRole)).toBeTruthy();
    
        fireEvent.click(getByRole(downloadButtonRole));

        expect(s3Spy).toHaveBeenCalledWith("1234.png");
        expect(await findByText("Download")).toBeInTheDocument();
    });
});

