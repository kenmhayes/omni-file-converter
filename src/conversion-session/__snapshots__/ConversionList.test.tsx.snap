// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConversionList renders a conversion list 1`] = `
<div
  className="container"
>
  <div
    className="card"
  >
    <div
      className="card-header"
    >
      <h4>
        Your converted files:
      </h4>
    </div>
    <div
      className="card-body"
    >
      <table
        className="table"
      >
        <thead>
          <tr>
            <th>
              File Name
            </th>
            <th>
              Converted File Type
            </th>
            <th>
              Status
            </th>
            <th>
              Download Link
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              test1.jpg
            </td>
            <td>
              png
            </td>
            <td>
              Ready to download
            </td>
            <td>
              <button
                className="btn btn-primary"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                Get Download Link
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
